### PR TITLE
Request composer

### PR DESCRIPTION
**Pull request by mcohoon**
Hi Philip,
I want to initiate the pull request process so you can start to comment and make any suggestions. There are probably still a few more features, that we might want to finalize before we do the final merge.

For Others, 
This pull request is to introduce the "Request Composer" into the Demos & Tools site. It is a simple query explorer like application which makes use of some of the more advanced features of Analytics Reporting API V4.

-Matt

### CODE REVIEW

**Comment by philipwalton** [comment]
I'm going to start taking a look at this, but before I make comments on code style, can you fix the issues reported by running `gulp lint`?

File: package.json [hunk 1]
```diff
@@ -12,23 +12,25 @@
...
-    "babel-core": "^6.11.4",
-    "babel-eslint": "^6.1.2",
-    "babel-loader": "^6.2.4",
-    "babel-polyfill": "^6.9.1",
-    "babel-preset-es2015": "^6.9.0",
-    "babel-preset-react": "^6.11.1",
-    "babel-preset-stage-0": "^6.5.0",
-    "camelcase": "^3.0.0",
-    "core-js": "^2.4.1",
+    "babel-core": "^6.1.13",
```
Comment (philipwalton) [inline package.json]: It looks like a lot of these have gone backwards. Can you make sure to include the latest changes from master?

File: package.json [hunk 2]
```diff
@@ -12,23 +12,25 @@
...
+    "eslint": "^2.8.0",
+    "eslint-plugin-babel": "^3.0.0",
+    "eslint-plugin-react": "^4.3.0",
+    "fixed-data-table-2": "^0.7.1",
+    "fs-extra": "^0.27.0",
+    "glob": "^7.0.3",
+    "gulp": "^3.9.0",
     "gulp-autoprefixer": "3.1.0",
     "gulp-concat": "^2.6.0",
     "gulp-cssnext": "^1.0.1",
```
Comment (philipwalton) [inline package.json]: I think `moment` is missing here. I got a build error when trying to run the site. If you delete your `node_modules` folder and run `npm install` again, you'll be able to see if anything's missing.

File: package.json (hunk 1 above)
Comment (mcohoon) [inline package.json]: [Fixed](https://github.com/mcohoon/ga-dev-tools/commit/010c6a539734b84a22480b3ddbc8cdb95bd871f8).

File: src/javascript/request-composer/components/request-composer.js [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+                  name="ReportRequest.FIELDS.segments" />
+              </div>
+              <div className="FormControl-info">
+                <label>
+                  <input
+                    className="Checkbox"
+                    type="checkbox"
+                    onChange={this.handleSegmentDefinitionToggle}
+                    checked={settings.useDefinition} />
+                  Show segment definitions instead of viewId.
```
Comment (philipwalton) [inline src/javascript/request-composer/components/request-composer.js]: This should say: "Show segment definitions instead of IDs."

File: meta.yaml [hunk 4]
```diff
@@ -74,6 +74,10 @@ projects:
   type: tool
   slug: query-explorer
   tech: [core_reporting_api, embed_api, management_api, metadata_api]
+- title: Request Composer
+  type: tool
+  slug: request-composer
```
Comment (philipwalton) [inline meta.yaml]: You can add a `status` key here so this shows up with a "New!" badge in the sidebar.

File: src/javascript/request-composer/store.js [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+ */
+function getInitalQueryParamsAndUpdateUrl() {
+
+  let defaultParams = {
+    'startDate': '30daysAgo',
+    'endDate': 'yesterday',
+    'cohortSize': 'Day',
+    'cohortMetrics': 'ga:cohortActiveUsers'
+  };
+  let storedParams = db.get('request-composer:params');
```
Comment (philipwalton) [inline src/javascript/request-composer/store.js]: Question: do you think it makes sense to store the params for each of the report types separately? I.e. if you switch tabs, change the metrics, and then switch back, do you expect your previously selected metrics for that report type to reappear?

File: src/javascript/request-composer/store.js [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+  // Don't assume that the presence any query params means it's a Query
+  // Explorer URL. Only use the query params if they exist and contain at least
+  // a metric value.
+  if (urlParams && urlParams['metrics']) {
+
+    // Some of the Query Explorer links out in the wild have double-encoded
+    // URL params. Check for the substring '%253A' (a colon, double-encoded),
+    // and if found then double-decode all params.
+    if (urlParams['metrics'].indexOf('%253A')) {
```
Comment (philipwalton) [inline src/javascript/request-composer/store.js]: This logic is only needed to handle old query explorer links, so it can be removed.

File: src/javascript/request-composer/store.js [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+      urlParams = mapValues(urlParams, (value) => decodeURIComponent(value));
+    }
+
+    // Remove the query params in the URL to prevent losing state on refresh.
+    // https://github.com/googleanalytics/ga-dev-tools/issues/61
+    if (history && history.replaceState) {
```
Comment (philipwalton) [inline src/javascript/request-composer/store.js]: This logic here from the Query Explorer was to support adding params to the URL query for sharing. Is this supported in the Request Composer?

File: src/javascript/request-composer/store.js [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+});
+
+
+// TODO(philipwalton): create middleware to save the params and settings
+// to localStorage.
+store.subscribe(function() {
+  let {params, settings} = store.getState();
+
+  db.set('request-composer:settings', settings);
+  gaAll('set', 'dimension3', qs.stringify(settings));
```
Comment (philipwalton) [inline src/javascript/request-composer/store.js]: This dimension is specifically for the Query Explorer. You can make your own for this tool or just remove this line.

File: src/javascript/request-composer/store.js [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+ * Gets the Requet Composer settings stored in local storage. If no settings
+ * exist, an empty object is returned.
+ * @return {Object} The settings object.
+ */
+function getDefaultSettingsAndUpdateTracker() {
+  let settings = db.get('request-composer:settings') || {
+    'requestType': 'HISTOGRAM'
+  };
+
+  gaAll('set', 'dimension3', qs.stringify(settings));
```
Comment (philipwalton) [inline src/javascript/request-composer/store.js]: This dimension is specifically for the Query Explorer. You can make your own for this tool or just remove this line.

File: src/javascript/embed-api/components/view-selector2.js [hunk 10]
```diff
@@ -235,6 +235,7 @@ gapi.analytics.ready(function() {
         account: this.account,
         property: this.property,
         view: this.view,
+        viewId: this.view.id,
```
Comment (philipwalton) [inline src/javascript/embed-api/components/view-selector2.js]: You can already get the view ID from the view object, why is this needed?

File: src/javascript/request-composer/components/request-composer.js [hunk 11]
```diff
@@ -0,0 +1,614 @@
...
+    let formActionClass = formControlClass + ' FormControl--action';
+    let requiredFormControlClass = formControlClass +' FormControl--required';
+    let maximumSelectionSize = 2;
+
+    return (
+      <div>
+        <h3 className="H3--underline">Select a view</h3>
+
+        <ViewSelector
+          viewId={params.viewId}
```
Comment (philipwalton) [inline src/javascript/request-composer/components/request-composer.js]: This isn't actually doing anything since the `ViewSelector` component isn't currently configured to initialize with a `viewId` prop. You could add it though by updating this logic: https://github.com/mcohoon/ga-dev-tools/blob/request-composer/src/javascript/components/view-selector.js#L62-L64

File: src/javascript/request-composer/request.js [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+    } else {
+      return request;
+    }
+    for (let i in dimensions) {
+      let dimension = {'name': dimensions[i]};
+      if (settings.requestType &&
+          settings.requestType == 'HISTOGRAM' &&
+          params.buckets) {
+        let histogramBuckets = [];
+        let buckets = params.buckets.split(/[ ,]+/);
```
Comment (philipwalton) [inline src/javascript/request-composer/request.js]: Nit: splitting on `/[ ,]+/` would allow the input to be something like "0,,,10,,,20" or "0 10 20". You could do `/\s*,\s*/` if you wanted to force a single comma.

File: src/javascript/request-composer/request.js [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+  return request;
+}
+
+function applyMetrics(request, params, settings) {
+  if (params.metrics &&
+      settings.requestType != 'COHORT') {
+    request.metrics = [];
+
+    let metrics = params.metrics.split(',');
+    for (let i in metrics) {
```
Comment (philipwalton) [inline src/javascript/request-composer/request.js]: You should use [for...of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of) loops here (and everywhere else).

```js
for (let metric of metrics) {
  request.metrics.push({expression: metric});
}
```

For..in loops return the key instead of the value and have issues where they also iterate over properties up the prototype chain.

File: src/javascript/request-composer/request.js [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+  }
+  return reportRequest;
+}
+
+function applyDateRanges(request, params, settings) {
+  if (params.startDate &&
+      params.endDate &&
+      settings.requestType != 'COHORT') {
+    request.dateRanges = [{
+      'startDate': params.startDate,
```
Comment (philipwalton) [inline src/javascript/request-composer/request.js]: According the the style guide, object keys shouldn't use quotes unless they need to (e.g. they contain a space or hyphen). Can you fix this here and elsewhere?

I'll see if I can add this to the lint rules.

File: src/javascript/request-composer/request.js [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import moment from 'moment';
+
+function buildReportRequest(params) {
```
Comment (philipwalton) [inline src/javascript/request-composer/request.js]: Can you add [JS doc](http://usejsdoc.org/) comments to all the functions/methods in this PR?

I've been a bit lax about it in some of my existing code, but I think we should standardize on enforcing it. It'll definitely make code reviews easier.

I'll add it to the linter and make sure all the other code in this repo passes.

File: src/javascript/request-composer/components/request-composer.js [hunk 16]
```diff
@@ -0,0 +1,614 @@
...
+        <form onSubmit={this.handleSubmit}>
+
+          <div className={requiredFormControlClass}>
+            <label className="FormControl-label">viewId</label>
+            <div className="FormControl-body">
+              <div className="FlexLine">
+                <input
+                  className="FormField FormFieldCombo-field"
+                  name="viewId"
+                  value={params.viewId}
```
Comment (philipwalton) [inline src/javascript/request-composer/components/request-composer.js]: I'm getting the following error that's due to these `value` props not being defined:

> Warning: RequestComposer is changing an uncontrolled input of type undefined to be controlled. Input elements should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled input element for the lifetime of the component. More info: https://fb.me/react-controlled-components

In the Query Explorer I use `someValue || ''` everywhere. You could do that or use the `defaultValue` prop. I'm not sure which is more idiomatic.

File: src/javascript/request-composer/components/request-composer.js [hunk 17]
```diff
@@ -0,0 +1,614 @@
...
+              </div>
+            </div>
+          </div>
+          ) :
+          null}
+
+
+          {settings.requestType == 'HISTOGRAM' ? (
+          <div className={formControlClass}>
+            <label className="FormControl-label">Histogram dimension</label>
```
Comment (philipwalton) [inline src/javascript/request-composer/components/request-composer.js]: A lot of these parameters aren't shown as being required, and if I try to make requests without them being present, I get JS errors instead of UI errors. Can you make sure all required fields are visually indicated as such via the `requiredFormControlClass` `className` prop.

File: src/javascript/embed-api/components/view-selector2.js (hunk 10 above)
Comment (mcohoon) [inline src/javascript/embed-api/components/view-selector2.js]: Resolved.

**Comment by philipwalton** [comment]
FYI, I'm not sure what's causing this, but the table generated by the Pivot Request does not allow for horizontal scrolling for me. I can help investigate this, but I suspect it's a problem with the component, and we may have to use a different one.

File: meta.yaml (hunk 4 above)
Comment (mcohoon) [inline meta.yaml]: Resolved.

File: src/javascript/request-composer/store.js (hunk 6 above)
Comment (mcohoon) [inline src/javascript/request-composer/store.js]: Resolved.

File: src/javascript/request-composer/store.js (hunk 7 above)
Comment (mcohoon) [inline src/javascript/request-composer/store.js]: Removed (for the time being).

File: src/javascript/request-composer/store.js (hunk 8 above)
Comment (mcohoon) [inline src/javascript/request-composer/store.js]: removed.

File: src/javascript/request-composer/store.js (hunk 9 above)
Comment (mcohoon) [inline src/javascript/request-composer/store.js]: removed.

File: src/javascript/embed-api/components/view-selector2.js (hunk 10 above)
Comment (mcohoon) [inline src/javascript/embed-api/components/view-selector2.js]: I think I was just mirroring your convenience variable `ids`. I have removed it.

File: src/javascript/request-composer/components/request-composer.js (hunk 11 above)
Comment (mcohoon) [inline src/javascript/request-composer/components/request-composer.js]: Fixed.

File: src/javascript/request-composer/request.js (hunk 12 above)
Comment (mcohoon) [inline src/javascript/request-composer/request.js]: Fixed.

File: src/javascript/request-composer/components/request-composer.js (hunk 17 above)
Comment (mcohoon) [inline src/javascript/request-composer/components/request-composer.js]: Fixed.

File: src/javascript/request-composer/components/request-composer.js (hunk 16 above)
Comment (mcohoon) [inline src/javascript/request-composer/components/request-composer.js]: Added `someValue || ''`

File: src/javascript/request-composer/request.js (hunk 15 above)
Comment (mcohoon) [inline src/javascript/request-composer/request.js]: Done.

File: src/javascript/request-composer/request.js (hunk 14 above)
Comment (mcohoon) [inline src/javascript/request-composer/request.js]: Done.

File: src/javascript/request-composer/store.js (hunk 5 above)
Comment (mcohoon) [inline src/javascript/request-composer/store.js]: I think I was trying to balance the repeating of the select2 code required to have multiple selectors, having shared parameters between the request types, and not permitting inappropriate dimensions and metrics for request types that they don't make sense. Perhaps if we updated the select2 code to parameterize some of the choices of metrics and dimensions, each query type could have its own dimensions and metrics. But it wouldn't make sense for each tab to have its own `viewId`, and the consistency between the tabs make sense. 

The current behavior appears match what you describe, but if you see something that gets dropped between sessions or tabs let me know.

Comment (philipwalton) [inline src/javascript/request-composer/store.js]: SGTM

**Comment by philipwalton** [comment]
@mcohoon, since Github recently added the ability to update a PR's base branch, I'm going to merge this into the `request-composer` branch in the main repo. I'll then add my changes as a PR to that branch, so you can review.
